pkg/proctree/events_processing.go

```diff
@@ -0,0 +1,50 @@
+package proctree
+
+import (
+	"github.com/aquasecurity/tracee/pkg/events"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+// ProcessEvent update the process tree according to arriving event
+func (tree *ProcessTree) ProcessEvent(traceeEvent *trace.Event) error {
+	switch events.ID(traceeEvent.EventID) {
+	case events.SchedProcessFork:
+		return tree.ProcessForkEvent(traceeEvent)
+	case events.SchedProcessExec:
+		return tree.ProcessExecEvent(traceeEvent)
+	case events.SchedProcessExit:
+		return tree.ProcessExitEvent(traceeEvent)
+	case events.Exit, events.InitNamespaces, events.HiddenKernelModule:
+		return nil
+	default:
+		return tree.processGeneralEvent(traceeEvent)
+	}
+}
+
+// processGeneralEvent fills process information with the data which resides in every event.
+// Warning: there are events with fake process information, because they are originated from
+// user mode. DO NOT call this function with these events, as it will corrupt the process tree.
+func (tree *ProcessTree) processGeneralEvent(event *trace.Event) error {
+	process, err := tree.getProcess(event.HostProcessID)
+	if err != nil {
+		process, err = tree.addGeneralEventProcess(event)
+		if err != nil {
+			return err
+		}
+	}
+	process.setGenInfoFromEventProtected(event)
+	_, err = tree.addGeneralEventThread(event)
+	if err != nil {
+		return err
+	}
+	process.mutex.RLock()
+	parent := process.getParent()
+	process.mutex.RUnlock()
+	if parent == nil {
+		_, err = tree.generateParentProcess(event.HostParentProcessID, event.ParentProcessID, process)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
```

Comment:
So if you will go into details, all the functions here are very basic and almost lockless.
For the get function of the processes and thread, we have to read-lock the mutexes of the map, but all the other things are optimized.
We use the `setGenInfoFromEventProtected` to initialized process information only if missing, without using locks (uses `sync.Once` object).
The `addGeneralEventThread` only create new thread if the thread is missing, else does nothing.
And again, we create a parent only if it is missing.

Reply:
@yanivagman notice that we currently only use the three basich `schedX` events, so this is not happening for all Tracee's events.

Reply:
> @yanivagman notice that we currently only use the three basich `schedX` events, so this is not happening for all Tracee's events.

But I see that you call this function in the default case as well (for any other event), isn't it right?

Reply:
I do, but it is not called in the pipeline right now.
When and if we put it in the pipeline for other events, than we can optimize it more and maybe filter some of the events.

Reply:
Ok I see. I thought the function is being used, but now I understand it is not.
So right now if you missed one of exec/fork/exit events then process tree state will never be updated for the relevant process, right? I thought this mechanism was used to resolve such cases

Reply:
So that was the general idea.
However, as we want to move control plane events we will receive much less lost events.
In addition, to support all processes in the process event stage is a bit of a trouble, as we use a map right now and I only signed `sched_process_exec`, `sched_process_fork` and `sched_process_exit`. If we sign each event, than we can just sign all non-problematic events to the general processing function. Maybe we should export it?
Anyways, I propose that we start like this, and in later PR we can add more events to the process tree processing.